### PR TITLE
[WIP] Ignore members in Encode.Auto and check optional members in Decode.Auto

### PR DESCRIPTION
**Pull request by alfonsogarciacaro**
I haven't completed the .NET part yet, what do you think of this approach?

### CODE REVIEW

File: src/Thoth.Json/Decode.fs [hunk 1]
```diff
@@ -805,17 +805,24 @@ module Decode =
...
-            (decoders, Ok []) ||> Array.foldBack (fun (name, decoder) acc ->
+            (decoders, Ok []) ||> Array.foldBack (fun (name, isOptional, decoder) acc ->
                 match acc with
                 | Error _ -> acc
                 | Ok result ->
-                    // TODO!!! Optional types shouldn't be required
-                    field name decoder path value
-                    |> Result.map (fun v -> v::result))
+                    if isOptional then
+                        // TODO: Copied from IOptionalGetter.Field. Should we unify it?
```
Comment (MangelMaxime) [inline src/Thoth.Json/Decode.fs]: If we can unify it using a function or an inline function I am for it.

We need to mark the function as internal or private so it is not exposed to users

File: src/Thoth.Json/Encode.fs [hunk 2]
```diff
@@ -338,7 +338,17 @@ module Encode =
         Option.map encoder >> Option.defaultWith (fun _ -> nil)
 
     type Auto =
-        static member toString(space : int, value : obj, ?forceCamelCase : bool) : string =
+        static member toString(space : int, value : obj, ?forceCamelCase : bool, ?ignoreMembers: string seq) : string =
```
Comment (MangelMaxime) [inline src/Thoth.Json/Encode.fs]: What do you think of renaming `ignoreMembers` to `ignoreProperties`?

For me, members are more when working with class and not record. But I can be wrong.
Also, I believe, `property` is the standard name for JSON object.

**Comment by MangelMaxime** [comment]
Do you have a use case in mind for `ignoreMembers`?

One use case I have in mind is that you have your whole model state and you don't want to send everything to the server only the properties containing data and not the properties which store the UI state.

We should also take some time to consider this fact. The `ignoreMembers` list is string based so there can be typos and compiler will not be able to detect them for you. 
Where if in the scenario I mentioned, if we split the model into 2 parts one for the UI state and one for the data then we can just send the data record. 

I am just doing the evil advocate so we can discuss the pro and cons of the feature :)

PS: I guess you used a `string list` because attributes data isn't kept at runtime?

**Comment by alfonsogarciacaro** [comment]
Yes, that's the use case I was thinking of. Well, it's a string list but it's the same if you write `Encode.object` manually so... We could do some tricks in Fable to avoid the strings, but it would be difficult to reproduce them on the .NET side.

Indeed the attributes are not kept in Fable runtime. Personally I also don't like polluting the model with this kind of attributes, and it goes a bit against the Thoth.Json purpose of just using functions instead of relying too much on metadata.

**Comment by MangelMaxime** [comment]
No worry I am not a fan of attributes neither :)

> but it's the same if you write `Encode.object` manually so...

I don't agree. When writing the `Encode.object` manually you still have the compiler having your back.

```fs
Encode.object [
    "refreshToken", Encode.string user.RefreshToken
]
```

If `user` doesn't have a property `RefreshToken` you can't compile your code. 

> We could do some tricks in Fable to avoid the strings, but it would be difficult to reproduce them on the .NET side.

No need for the tricks no worries :). 

One more point.
If we add this feature, then it's a one-way conversion. You can't (I think) build the model back from the generated JSON there will be some missing required fields except if all are optional. We will need to make it explicitly in the docs.

**Comment by alfonsogarciacaro** [comment]
Ok, you convince me. Then maybe, for symmetry with `Decode.Auto`, we can just automatically remove `None` fields from the JSON. So users can do something like:

```fsharp
{ myType with IgnoreProp = None } |> Encode.Auto.toString
```

Though this may imply either using reflection also in `Encode.Auto`, or just removing all `null` fields from the JSON.

**Comment by alfonsogarciacaro** [comment]
Ok, I've updated this so None values are automatically excluded from the JSON object. The only quirk is this will happen with null values too, but it's probably fine as missing fields will be interpreted as null by the auto decoder and we should avoid null anyways :)

**Comment by MangelMaxime** [comment]
Is this PR still needed?

**Comment by alfonsogarciacaro** [comment]
TBH, I don't remember. But I can't see the "Optional members are ignored" test in the latest test suite so I think this feature is not implemented yet. If we still want it we probably need to recycle this PR.

**Comment by MangelMaxime** [comment]
Let's close this and see if people actually need it.

In the beta version of Thoth.Json they can mix auto-coders and manual coders in order to ignore members.

**Pull request closed without merging**